zenoh/src/net/routing/interceptor/authorization.rs

```diff
@@ -32,7 +35,81 @@ use zenoh_util::net::get_interface_names_by_addr;
 type PolicyForSubject = FlowPolicy;
 
 type PolicyMap = HashMap<usize, PolicyForSubject, RandomState>;
-type SubjectMap = HashMap<Subject, usize, RandomState>;
+
+#[derive(Debug, Clone)]
+pub(crate) struct SubjectMap {
+    inner: TrieMap<Subject, usize>,
+}
+
+impl SubjectMap {
+    pub(crate) fn builder() -> SubjectMapBuilder {
+        SubjectMapBuilder::new()
+    }
+
+    pub(crate) fn get(&self, subjects: &[Subject]) -> Vec<(Vec<Subject>, &usize)> {
+        let mut subjects = subjects.to_owned();
+        subjects.sort_unstable();
+        let mut res: Vec<(Vec<Subject>, &usize)> = vec![];
+        for subject in subjects {
+            let map_query: Vec<(Vec<Subject>, &usize)> =
+                self.inner.predictive_search([subject]).collect();
+            for (subject, id) in map_query {
+                if subject.len() > 1 {
+                    for s in &subject {
+                        if subject.binary_search(s).is_err() {
+                            continue;
+                        }
+                    }
+                }
+                res.push((subject, id))
+            }
+        }
+        res
+    }
+}
+
+impl Default for SubjectMap {
+    fn default() -> Self {
+        Self {
+            inner: SubjectMapBuilder::new().builder.build(),
+        }
+    }
+}
+
+pub(crate) struct SubjectMapBuilder {
+    builder: TrieMapBuilder<Subject, usize>,
+    hashmap: HashMap<Vec<Subject>, usize, RandomState>, // builder does not provide "insert_or_get" function, so we must use a hashmap during construction
+    id_counter: usize,
+}
+
+impl SubjectMapBuilder {
+    pub(crate) fn new() -> Self {
+        Self {
+            builder: TrieMapBuilder::new(),
+            hashmap: HashMap::with_hasher(RandomState::default()),
+            id_counter: 0,
+        }
+    }
+
+    pub(crate) fn build(self) -> SubjectMap {
+        SubjectMap {
+            inner: self.builder.build(),
+        }
+    }
+
+    /// Assumes subject contains at most one instance of each Subject variant
+    pub(crate) fn insert_or_get(&mut self, subject: Vec<Subject>) -> usize {
```

Comment:
It's not clear to me why a builder should have an `insert_or_get`. Usually a builder always overwrite existing values. E.g.:
```rust
let a = mybuilder()
    .foo("abc")
    .foo("def")
    .build();
    
assert_eq!(a.foo(), "def");
```

Reply:
This is required for the subject map because we want to avoid duplicating a subject combination. If the name `insert_or_get` is not fitting what should we call it?

Reply:
It's unclear to me then who is providing the `Vec<Subject>` as parameter that is ignored in case a subject is already present. What I want to be sure of is that a misusage of `insert_or_get` can lead to a security misconfiguration.

Reply:
The `Vec<Subject>` is the "subject", or better say the subject combination. Each individual subject combination is mapped to an id in the SubjectMap (the combination containing at lease one subject).
The issue is that when parsing the ACL config, we perform the cartesian product of `interfaces`, `cert_common_names` and `usernames` within each configured subject, to derive the subject combinations from each individual subject object in config, which can yield subject combinations that intersect with another subject object in the config. Depending on the parsing order, some combinations may have already been mapped, and overwriting those mappings would require updating previously parsed subject references in the intermediate data structures.
With this context, could you please explain what security misconfiguration can arise?

Reply:
It's still not clear to me how the linear parsing of the configuration could lead lead to different parsing order. Would you mind clarifying this point?

Reply:
Sorry I may have misspoken, I'll reiterate my point.

What I mean is that depending on the subjects declared in the config (in the `subjects` array), we may encounter during config parsing a subject combination that was previously put in the subject map and associated to another subject id (the unique string identifier provided in the config). The goal of `insert_or_get` is to avoid overwriting previously mapped subject combinations, in order to avoid updating the reference in the parsing data structure for previously parsed subject config (the data structure being a hashmap that maps subject string Ids to a `Vec<usize>` containing the "references" to the subject combinations).

On another note, I realize now that calling "subjects" both the combinations and the individual characteristics (interfaces, cert_common_names and usernames) is confusing.